streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/LegacySubscriptionInfoSerde.java

```diff
@@ -0,0 +1,276 @@
+
+            buf.rewind();
+
+            return buf;
+        } else if (usedVersion == 2) {
+            final byte[] endPointBytes = prepareUserEndPoint(this.userEndPoint);
+
+            final ByteBuffer buf = ByteBuffer.allocate(
+                4 + // version
+                    16 + // client ID
+                    4 + prevTasks.size() * 8 + // length + prev tasks
+                    4 + standbyTasks.size() * 8 + // length + standby tasks
+                    4 + endPointBytes.length
+            );
+
+            buf.putInt(2); // version
+            encodeClientUUID(buf, processId());
+            encodeTasks(buf, prevTasks);
+            encodeTasks(buf, standbyTasks);
+            encodeUserEndPoint(buf, endPointBytes);
+
+            buf.rewind();
+
+            return buf;
+        } else if (usedVersion == 1) {
+            final ByteBuffer buf1 = ByteBuffer.allocate(
+                4 + // version
+                    16 + // client ID
+                    4 + prevTasks.size() * 8 + // length + prev tasks
+                    4 + standbyTasks.size() * 8
+            );
+
+            buf1.putInt(1); // version
+            encodeClientUUID(buf1, processId());
+            encodeTasks(buf1, prevTasks);
+            encodeTasks(buf1, standbyTasks);
+            buf1.rewind();
+            return buf1;
+        } else {
+            throw new IllegalStateException("Unknown metadata version: " + usedVersion
+                                                + "; latest supported version: " + LATEST_SUPPORTED_VERSION);
+        }
+    }
+
+    public static void encodeClientUUID(final ByteBuffer buf, final UUID processId) {
+        buf.putLong(processId.getMostSignificantBits());
+        buf.putLong(processId.getLeastSignificantBits());
+    }
+
+    public static void encodeTasks(final ByteBuffer buf,
+                                   final Collection<TaskId> taskIds) {
+        buf.putInt(taskIds.size());
+        for (final TaskId id : taskIds) {
+            id.writeTo(buf);
+        }
+    }
+
+    public static void encodeUserEndPoint(final ByteBuffer buf,
+                                             final byte[] endPointBytes) {
+        if (endPointBytes != null) {
+            buf.putInt(endPointBytes.length);
+            buf.put(endPointBytes);
+        }
+    }
+
+    public static byte[] prepareUserEndPoint(final String userEndPoint) {
+        if (userEndPoint == null) {
+            return new byte[0];
+        } else {
+            return userEndPoint.getBytes(StandardCharsets.UTF_8);
+        }
+    }
+
+    /**
+     * @throws TaskAssignmentException if method fails to decode the data
+     */
+    public static LegacySubscriptionInfoSerde decode(final ByteBuffer data) {
+
+        // ensure we are at the beginning of the ByteBuffer
+        data.rewind();
+
+        final int usedVersion = data.getInt();
+        if (usedVersion == 4 || usedVersion == 3) {
```

Comment:
Do you think a switch block is easier to interpret?

Reply:
Maybe? I don't know how to determine an answer to this question.

If you like, I can change it to a switch block in a later PR, but I'm reluctant to roll the dice _again_ on the integration tests if there is no other feedback requiring changes.